src/main/scala/it/pagopa/interop/agreementprocess/api/impl/AgreementApiServiceImpl.scala

```diff
@@ -1,418 +1,517 @@
+    eService: CatalogManagement.EService,
+    consumer: TenantManagement.Tenant,
+    requesterOrgId: UUID
+  )(implicit contexts: Seq[(String, String)]): Future[AgreementManagement.Agreement] = {
+    val nextStateByAttributes = AgreementStateByAttributesFSM.nextState(agreement.state, eService, consumer)
+    val suspendedByConsumer   = suspendedByConsumerFlag(agreement, requesterOrgId, AgreementState.ACTIVE)
+    val suspendedByProducer   = suspendedByProducerFlag(agreement, requesterOrgId, AgreementState.ACTIVE)
+    val suspendedByPlatform   = suspendedByPlatformFlag(nextStateByAttributes)
+    val newState              =
+      agreementStateByFlags(nextStateByAttributes, suspendedByProducer, suspendedByConsumer, suspendedByPlatform)
+    val firstActivation       =
+      agreement.state == AgreementManagement.AgreementState.PENDING && newState == AgreementManagement.AgreementState.ACTIVE
+    val updateSeed            =
+      if (firstActivation)
+        AgreementManagement.UpdateAgreementSeed(
+          state = newState,
+          certifiedAttributes = matchingCertifiedAttributes(eService, consumer),
+          declaredAttributes = matchingDeclaredAttributes(eService, consumer),
+          verifiedAttributes = matchingVerifiedAttributes(eService, consumer),
+          suspendedByConsumer = suspendedByConsumer,
+          suspendedByProducer = suspendedByProducer,
+          suspendedByPlatform = suspendedByPlatform
+        )
+      else
+        AgreementManagement.UpdateAgreementSeed(
+          state = newState,
+          certifiedAttributes = agreement.certifiedAttributes,
+          declaredAttributes = agreement.declaredAttributes,
+          verifiedAttributes = agreement.verifiedAttributes,
+          suspendedByConsumer = suspendedByConsumer,
+          suspendedByProducer = suspendedByProducer,
+          suspendedByPlatform = suspendedByPlatform
         )
-      ),
-      state = agreementStateToApi(agreement.state),
-      attributes = attributes,
-      suspendedByConsumer = agreement.suspendedByConsumer,
-      suspendedByProducer = agreement.suspendedByProducer
+
+    val failureStates = List(
+      AgreementManagement.AgreementState.DRAFT,
+      AgreementManagement.AgreementState.PENDING,
+      AgreementManagement.AgreementState.MISSING_CERTIFIED_ATTRIBUTES
     )
+
+    def failOnActivationFailure() = Future
+      .failed(AgreementActivationFailed(agreement.id))
+      .whenA(failureStates.contains(newState))
+
+    for {
+      updated <- agreementManagementService.updateAgreement(agreement.id, updateSeed)
+      _       <- failOnActivationFailure()
+      _       <- authorizationManagementService
+        .updateStateOnClients(
+          eServiceId = agreement.eserviceId,
+          consumerId = agreement.consumerId,
+          agreementId = agreement.id,
+          state = toClientState(newState)
+        )
+        .unlessA(failureStates.contains(newState))
+    } yield updated
   }
```

Comment:
There's really no need for a for comprehension if 2/3 of the results are ignored, consider using `>>` or `<*` or use `unlessA` (or `whenA`) like in the last statement.